File: 백준/Silver/12101. 1， 2， 3 더하기 2/1， 2， 3 더하기 2.py
def dfs(depth):
    global cnt
    global stop
    
    if stop: # k번째 찾았으면 종료
        return
    
    if len(arr) > n: # 1을 n개 까지만 뽑도록
        return
    
    if sum(arr) > n: # 뽑은 수 합이 이미 n을 넘었으면 종료
        return
    
    if sum(arr) == n: # 뽑은 수 합이 n이면 카운트
        cnt += 1
    
    if cnt == k: # k번째 찾았으면 종료하도록 stop -> True
        print("+".join(map(str, arr)))
        stop = True
        return
        
    for i in range(1, 4):
        arr.append(i)
        dfs(depth+1)
        arr.pop()


n, k = map(int, input().split())
cnt = 0
arr = []
stop = False
dfs(1)

if not stop: # 함수 종료 후에도 k번째가 없었으면 -1 출력
    print(-1)

File: 백준/Silver/12101. 1， 2， 3 더하기 2/1， 2， 3 더하기 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static int n, k, cnt;
    static boolean stop = false;
    static List<Integer> arr = new ArrayList<>();
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
        cnt = 0;

        backtrack(0);
        if (!stop) {
            System.out.println(-1);
        }
    }

    public static void backtrack(int sum) {
        if (stop) {
            return;
        }

        if (sum > n) { // 합이 n을 넘으면 종료
            return;
        }

        if (sum == n) {
            cnt++;
        }

        if (cnt == k) { // k번째 찾으면
            stop = true; // 종료 설정
            for (int i = 0; i < arr.size()-1; i++) {
                sb.append(arr.get(i)).append("+");
            }
            sb.append(arr.get(arr.size()-1));
            System.out.println(sb);
            return;
        }

        for (int i = 1; i < 4; i++) {
            arr.add(i);
            backtrack(sum+i);
            arr.remove(arr.size()-1);
        }
    }
}
